test/common/TestGradient.hs

```diff
@@ -344,15 +380,26 @@ adoptTests = testGroup "Tests of the port of adopt code"
 
 test_conv2d_dInp :: Assertion
 test_conv2d_dInp =
-  -- Below, @17.3@ is just @OS.constant 17.3@ via a @Num@ instance.
-  let arrK = 0 {-17.3-} :: OS.Array '[2, 2 {-6-}, 3, 4] Double
-      arrB = 0 {-2.28-} :: OS.Array '[5, 2, 6, 7] Double
-        -- this fails with the commented out values;
-        -- and fails to typecheck with the commented out dimension size,
-        -- because conv2d_dInp lacks the arrA parameter
-  in assertEqualUpToEpsilon 1e-7
-       (revDt (conv2d (constant arrK))  -- gradient wrt second argument
-              arrB   -- dummy, should be arrA, the point at which we take
-                     -- the gradient: https://en.wikipedia.org/wiki/Gradient
-              arrB )  -- sensitivity; should we normally test with 1?
-       (conv2d_dInp arrK arrB)  -- expected result
+  let -- Input of shape: batch x chas x height x width
+      arrA   = 1 :: OS.Array '[5, 2, 4, 8] Double
+      -- Filters of shape: num_filters x chas x kernel_height x kernel_width
+      arrK   = 1 :: OS.Array '[7, 2, 1, 3] Double
+      -- Output gradient of shape: batch x chas x output_height x output_width
+      arrB'  = 1 :: OS.Array '[5, 7, 4, 8] Double
+      -- Compare the ad version against the manual derivative.
+      dInp   = conv2d_dInp arrK arrB'
```

Comment:
I still can't wrap my head around why conv2d_dInp doesn't take arrA as an argument. Does it assume it's 0, 1 or some other fixed value? At what point is the gradient computed?

Reply:
Hi Mikolaj,
In the general case, ie for some arbitrary function f(x), we do need the point 'x' at which the gradient is defined. For example if I were to write a manual version of the gradient for f(x) = x^2:
```
sqrt x = x * x
sqrt_dInp x dOut = 2 * x * dOut -- the VJP of sqrt at x
```
But convolution is a linear function (in the sense of a linear combination) in all of its arguments: 'conv(input, filters)' depends on 'input' linearly and depends on 'filters' also linearly. We can see this in the definition for convolution at each index: it involves only a product of 'input' and 'filters' (then a sum).
A linear function has the same gradient at every point (eg f(x) = x ==> f'(x) = 1 for all x) so there is no need to take the arrA argument. This would not be true for so-called "non-linearities".
Hope this helps. We can chat more in the meeting!
Cheers
Tran

Reply:
Oh, great, thank you, I knew I would learn some useful maths by asking clueless questions just following syntax and types. That means I can ask QuickCheck to randomly generate arbitrary "point" arguments and test with them. If I knew this earlier, I could even test automatic derivative of convolution in isolation. I can't wait to apply QuickCheck to this code.